File: app/src/main/java/com/example/advancedprayertimes/ui/Fragments/SettingPreferenceFragments/PrayerBeginningEndSettingPreferencesFragment.java
package com.example.advancedprayertimes.UI.Fragments.SettingPreferenceFragments;

import static android.content.Context.MODE_PRIVATE;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.preference.ListPreference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceFragmentCompat;

import com.example.advancedprayertimes.Logic.AppEnvironment;
import com.example.advancedprayertimes.Logic.DataManagementUtil;
import com.example.advancedprayertimes.Logic.Entities.Setting_Entities.PrayerSettingsEntity;
import com.example.advancedprayertimes.Logic.Entities.Setting_Entities.PrayerTimeBeginningEndSettingsEntity;
import com.example.advancedprayertimes.Logic.Enums.EPrayerTimeMomentType;
import com.example.advancedprayertimes.Logic.Enums.EPrayerTimeType;
import com.example.advancedprayertimes.Logic.Enums.ESupportedAPIs;
import com.example.advancedprayertimes.R;
import com.google.gson.Gson;

import java.util.AbstractMap;
import java.util.stream.Stream;

public class PrayerBeginningEndSettingPreferencesFragment extends PreferenceFragmentCompat
{
    // region static fields

    private static String PREFERENCE_CATEGORY_API_SETTINGS = "apiSettingsPreferenceCategory";
    private static String PREFERENCE_NAME_API_SELECTION = "apiSelection";
    private static String PREFERENCE_NAME_MINUTE_ADJUSTMENT_SELECTION = "minuteAdjustmentSelection";
    private static String PREFERENCE_NAME_FAJR_DEGREE_SELECTION = "fajrCalculationDegree";
    private static String PREFERENCE_NAME_ISHA_DEGREE_SELECTION = "ishaCalculationDegree";

    // endregion static fields

    // region fields

    private SharedPreferences.OnSharedPreferenceChangeListener _preferenceChangeListener;

    private PreferenceCategory apiSettingsPreferenceCategory = null;
    private ListPreference apiSelectionListPreference = null;
    private ListPreference minuteAdjustmentListPreference = null;
    private ListPreference fajrDegreesListPreference = null;
    private ListPreference ishaDegreesListPreference = null;

    AbstractMap.SimpleEntry<EPrayerTimeType, EPrayerTimeMomentType> prayerTypeWithMomentType;
    private boolean _isBeginning;

    // endregion fields

    public PrayerBeginningEndSettingPreferencesFragment(EPrayerTimeType prayerType, boolean isBeginning)
    {
        super();

        _isBeginning = isBeginning;
        prayerTypeWithMomentType = new AbstractMap.SimpleEntry<EPrayerTimeType, EPrayerTimeMomentType>(prayerType, isBeginning ? EPrayerTimeMomentType.Beginning : EPrayerTimeMomentType.End);
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey)
    {
        setPreferencesFromResource(R.xml.prayer_beginning_end_settings_preferences, rootKey);

        apiSettingsPreferenceCategory = this.findPreference(PREFERENCE_CATEGORY_API_SETTINGS);

        apiSelectionListPreference = this.findPreference(PREFERENCE_NAME_API_SELECTION);
        minuteAdjustmentListPreference = this.findPreference(PREFERENCE_NAME_MINUTE_ADJUSTMENT_SELECTION);
        fajrDegreesListPreference = this.findPreference(PREFERENCE_NAME_FAJR_DEGREE_SELECTION);
        ishaDegreesListPreference = this.findPreference(PREFERENCE_NAME_ISHA_DEGREE_SELECTION);

        if (apiSelectionListPreference == null || minuteAdjustmentListPreference == null || fajrDegreesListPreference == null
                || ishaDegreesListPreference == null || apiSettingsPreferenceCategory == null)
        {
            return;
        }

        configureAPISelector(apiSelectionListPreference);

        ESupportedAPIs selectedAPI = ESupportedAPIs.Undefined;
        int minuteAdjustmentValue = 0;
        double fajrDegreeValue = -12.0;
        double ishaDegreeValue = -12.0;

        PrayerTimeBeginningEndSettingsEntity settings = AppEnvironment.prayerSettingsByPrayerType.get(prayerTypeWithMomentType.getKey()).GetBeginningEndSettingByMomentType(_isBeginning);

        if(settings == null)
        {
            settings = new PrayerTimeBeginningEndSettingsEntity(selectedAPI, minuteAdjustmentValue, fajrDegreeValue, ishaDegreeValue);
            AppEnvironment.prayerSettingsByPrayerType.get(prayerTypeWithMomentType.getKey()).SetBeginningEndSettingByMomentType(_isBeginning, settings);
        }
        else
        {
            selectedAPI = settings.get_api();
            minuteAdjustmentValue = settings.get_minuteAdjustment();

            if(settings.getFajrCalculationDegree() != null)
            {
                fajrDegreeValue = settings.getFajrCalculationDegree();
            }

            if(settings.getIshaCalculationDegree() != null)
            {
                ishaDegreeValue = settings.getIshaCalculationDegree();
            }
        }

        apiSelectionListPreference.setValue(selectedAPI.toString());
        minuteAdjustmentListPreference.setValue(String.valueOf(minuteAdjustmentValue));
        fajrDegreesListPreference.setValue(fajrDegreeValue + "");
        ishaDegreesListPreference.setValue(ishaDegreeValue + "");

        updatePreferenceVisibility();

        // change listener for all preference value changes
        _preferenceChangeListener = this::onPreferenceChange;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item)
    {
        return super.onOptionsItemSelected(item);
    }

    private void onPreferenceChange(SharedPreferences sharedPreferences, String key)
    {
        PrayerSettingsEntity prayerSettings = AppEnvironment.prayerSettingsByPrayerType.get(prayerTypeWithMomentType.getKey());

       if(prayerSettings == null)
        {
            return;
        }

       try
       {
           if(key.equals(apiSelectionListPreference.getKey()) && sharedPreferences.contains(apiSelectionListPreference.getKey()))
           {
               ESupportedAPIs api = ESupportedAPIs.valueOf(sharedPreferences.getString(apiSelectionListPreference.getKey(), null));
               prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning).set_api(api);

               updatePreferenceVisibility();
           }
           else if(key.equals(minuteAdjustmentListPreference.getKey()) && sharedPreferences.contains(minuteAdjustmentListPreference.getKey()))
           {
               int minuteAdjustment = sharedPreferences.getInt(minuteAdjustmentListPreference.getKey(), 0);
               prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning).set_minuteAdjustment(minuteAdjustment);
           }
           else if(key.equals(fajrDegreesListPreference.getKey()) && sharedPreferences.contains(fajrDegreesListPreference.getKey()))
           {
               Double fajrCalculationDegrees = (double) sharedPreferences.getFloat(fajrDegreesListPreference.getKey(), 0.0f);
               prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning).setFajrCalculationDegree(fajrCalculationDegrees);
           }
           else if(key.equals(ishaDegreesListPreference.getKey()) && sharedPreferences.contains(ishaDegreesListPreference.getKey()))
           {
               Double ishaCalculationDegrees = (double) sharedPreferences.getFloat(ishaDegreesListPreference.getKey(), 0.0f);
               prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning).setIshaCalculationDegree(ishaCalculationDegrees);
           }

           Gson gson = new Gson();

           String jsonString = gson.toJson(prayerSettings);
           this.getActivity().getSharedPreferences(AppEnvironment.GLOBAL_SHARED_PREFERENCE_NAME, MODE_PRIVATE).edit().putString(DataManagementUtil.GetTimeSettingsEntityKeyForSharedPreference(prayerTypeWithMomentType.getKey()), jsonString).commit();
       }
       catch(Exception e)
       {
            e.printStackTrace();
       }
    }

    // region methods

    private void configureAPISelector(ListPreference apiSelectionListPreference)
    {
        String[] apiNamesArray = null;

        if(prayerTypeWithMomentType.getKey() == EPrayerTimeType.Duha)
        {
            apiNamesArray = Stream.of(ESupportedAPIs.values()).filter(x -> x == ESupportedAPIs.Undefined || x == ESupportedAPIs.Muwaqqit).map(ESupportedAPIs::name).toArray(String[]::new);
        }
        else
        {
            apiNamesArray = Stream.of(ESupportedAPIs.values()).map(ESupportedAPIs::name).toArray(String[]::new);
        }

        apiSelectionListPreference.setEntries(apiNamesArray);
        apiSelectionListPreference.setEntryValues(apiNamesArray);
    }

    private void updatePreferenceVisibility()
    {
        boolean isDegreeAPISelected =
                ESupportedAPIs.valueOf(apiSelectionListPreference.getValue()) == ESupportedAPIs.Muwaqqit
                ||
                ESupportedAPIs.valueOf(apiSelectionListPreference.getValue()) == ESupportedAPIs.AlAdhan;

        boolean showFajrDegreeSelector = isDegreeAPISelected && PrayerTimeBeginningEndSettingsEntity.FAJR_DEGREE_TYPES.contains(prayerTypeWithMomentType);
        boolean showIshaDegreeSelector = isDegreeAPISelected && PrayerTimeBeginningEndSettingsEntity.ISHA_DEGREE_TYPES.contains(prayerTypeWithMomentType);

        fajrDegreesListPreference.setVisible(showFajrDegreeSelector);
        ishaDegreesListPreference.setVisible(showIshaDegreeSelector);

        apiSettingsPreferenceCategory.setVisible(fajrDegreesListPreference.isVisible() || ishaDegreesListPreference.isVisible());
    }

    // endregion methods

    @Override
    public void onPause()
    {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(_preferenceChangeListener);
    }

    @Override
    public void onResume()
    {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(_preferenceChangeListener);
    }
}

File: app/src/main/kotlin/com/example/advancedprayertimes/ui/fragments/PrayerBeginningEndSettingPreferencesFragment.kt
package com.example.advancedprayertimes.ui.fragments

import android.content.Context
import com.example.advancedprayertimes.logic.DataManagementUtil.GetTimeSettingsEntityKeyForSharedPreference
import com.example.advancedprayertimes.logic.enums.EPrayerTimeType
import androidx.preference.PreferenceFragmentCompat
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import com.example.advancedprayertimes.logic.enums.EPrayerTimeMomentType
import android.os.Bundle
import com.example.advancedprayertimes.R
import com.example.advancedprayertimes.logic.enums.ESupportedAPIs
import com.example.advancedprayertimes.logic.setting_entities.PrayerTimeBeginningEndSettingsEntity
import com.example.advancedprayertimes.logic.AppEnvironment
import android.content.SharedPreferences
import android.view.MenuItem
import androidx.preference.ListPreference
import androidx.preference.PreferenceCategory
import com.google.gson.Gson
import java.util.AbstractMap
import java.util.stream.Stream
import kotlin.streams.asSequence

class PrayerBeginningEndSettingPreferencesFragment(
    prayerType: EPrayerTimeType,
    private val _isBeginning: Boolean
) : PreferenceFragmentCompat() {

    private var _preferenceChangeListener: OnSharedPreferenceChangeListener? = null
    private var apiSettingsPreferenceCategory: PreferenceCategory? = null
    private var apiSelectionListPreference: ListPreference? = null
    private var minuteAdjustmentListPreference: ListPreference? = null
    private var fajrDegreesListPreference: ListPreference? = null
    private var ishaDegreesListPreference: ListPreference? = null
    var prayerTypeWithMomentType: AbstractMap.SimpleEntry<EPrayerTimeType, EPrayerTimeMomentType>

    override fun onCreatePreferences(savedInstanceState: Bundle, rootKey: String) {
        setPreferencesFromResource(R.xml.prayer_beginning_end_settings_preferences, rootKey)
        apiSettingsPreferenceCategory = findPreference(PREFERENCE_CATEGORY_API_SETTINGS)
        apiSelectionListPreference = findPreference(PREFERENCE_NAME_API_SELECTION)
        minuteAdjustmentListPreference = findPreference(PREFERENCE_NAME_MINUTE_ADJUSTMENT_SELECTION)
        fajrDegreesListPreference = findPreference(PREFERENCE_NAME_FAJR_DEGREE_SELECTION)
        ishaDegreesListPreference = findPreference(PREFERENCE_NAME_ISHA_DEGREE_SELECTION)
        if (apiSelectionListPreference == null || minuteAdjustmentListPreference == null || fajrDegreesListPreference == null || ishaDegreesListPreference == null || apiSettingsPreferenceCategory == null) {
            return
        }
        configureAPISelector(apiSelectionListPreference!!)
        var selectedAPI = ESupportedAPIs.Undefined
        var minuteAdjustmentValue = 0
        var fajrDegreeValue = -12.0
        var ishaDegreeValue = -12.0
        var settings = AppEnvironment.prayerSettingsByPrayerType[prayerTypeWithMomentType.key]!!
            .GetBeginningEndSettingByMomentType(_isBeginning)
        if (settings == null) {
            settings = PrayerTimeBeginningEndSettingsEntity(
                selectedAPI,
                minuteAdjustmentValue,
                fajrDegreeValue,
                ishaDegreeValue
            )
            AppEnvironment.prayerSettingsByPrayerType[prayerTypeWithMomentType.key]!!
                .SetBeginningEndSettingByMomentType(_isBeginning, settings)
        } else {
            selectedAPI = settings.api
            minuteAdjustmentValue = settings.minuteAdjustment
            if (settings.fajrCalculationDegree != null) {
                fajrDegreeValue = settings.fajrCalculationDegree!!
            }
            if (settings.ishaCalculationDegree != null) {
                ishaDegreeValue = settings.ishaCalculationDegree!!
            }
        }
        apiSelectionListPreference!!.value = selectedAPI.toString()
        minuteAdjustmentListPreference!!.value = minuteAdjustmentValue.toString()
        fajrDegreesListPreference!!.value = fajrDegreeValue.toString() + ""
        ishaDegreesListPreference!!.value = ishaDegreeValue.toString() + ""
        updatePreferenceVisibility()

        // change listener for all preference value changes
        _preferenceChangeListener =
            OnSharedPreferenceChangeListener { sharedPreferences: SharedPreferences, key: String ->
                onPreferenceChange(
                    sharedPreferences,
                    key
                )
            }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return super.onOptionsItemSelected(item)
    }

    private fun onPreferenceChange(sharedPreferences: SharedPreferences, key: String) {
        val prayerSettings = AppEnvironment.prayerSettingsByPrayerType[prayerTypeWithMomentType.key]
            ?: return
        if (key == apiSelectionListPreference!!.key && sharedPreferences.contains(
                apiSelectionListPreference!!.key
            )
        ) {
            val api = ESupportedAPIs.valueOf(apiSelectionListPreference!!.value)
            prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning)!!.api = api
            updatePreferenceVisibility()
        } else if (key == minuteAdjustmentListPreference!!.key && sharedPreferences.contains(
                minuteAdjustmentListPreference!!.key
            )
        ) {
            val minuteAdjustment = minuteAdjustmentListPreference!!.value.toInt()
            prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning)!!.minuteAdjustment =
                minuteAdjustment
        } else if (key == fajrDegreesListPreference!!.key && sharedPreferences.contains(
                fajrDegreesListPreference!!.key
            )
        ) {
            val fajrCalculationDegrees = fajrDegreesListPreference!!.value.toDouble()
            prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning)!!.fajrCalculationDegree =
                fajrCalculationDegrees
        } else if (key == ishaDegreesListPreference!!.key && sharedPreferences.contains(
                ishaDegreesListPreference!!.key
            )
        ) {
            val ishaCalculationDegrees = ishaDegreesListPreference!!.value.toDouble()
            prayerSettings.GetBeginningEndSettingByMomentType(_isBeginning)!!.ishaCalculationDegree =
                ishaCalculationDegrees
        }
        val jsonString = Gson().toJson(prayerSettings)
        val globalSharedPreference = this.requireActivity()
            .getSharedPreferences(
                AppEnvironment.GLOBAL_SHARED_PREFERENCE_NAME,
                Context.MODE_PRIVATE
            )
        globalSharedPreference.edit().putString(
            GetTimeSettingsEntityKeyForSharedPreference(prayerTypeWithMomentType.key),
            jsonString
        ).commit()
    }

    // region methods
    private fun configureAPISelector(apiSelectionListPreference: ListPreference) {
        var apiNamesArray: Array<String>? = null
        apiNamesArray = if (prayerTypeWithMomentType.key === EPrayerTimeType.Duha) {
            Stream.of(*ESupportedAPIs.values()).asSequence()
                .filter { x: ESupportedAPIs -> x === ESupportedAPIs.Undefined || x === ESupportedAPIs.Muwaqqit }
                .map { x -> x.toString() }
                .toList().toTypedArray()
        } else {
            Stream.of(*ESupportedAPIs.values()).asSequence()
                .map { obj: ESupportedAPIs -> obj.name }
                .toList().toTypedArray()
        }
        apiSelectionListPreference.entries = apiNamesArray
        apiSelectionListPreference.entryValues = apiNamesArray
    }

    private fun updatePreferenceVisibility() {
        val isDegreeAPISelected = (ESupportedAPIs.valueOf(
            apiSelectionListPreference!!.value
        ) === ESupportedAPIs.Muwaqqit
                ||
                ESupportedAPIs.valueOf(apiSelectionListPreference!!.value) === ESupportedAPIs.AlAdhan)
        val showFajrDegreeSelector =
            isDegreeAPISelected && PrayerTimeBeginningEndSettingsEntity.FAJR_DEGREE_TYPES.contains(
                prayerTypeWithMomentType
            )
        val showIshaDegreeSelector =
            isDegreeAPISelected && PrayerTimeBeginningEndSettingsEntity.ISHA_DEGREE_TYPES.contains(
                prayerTypeWithMomentType
            )
        fajrDegreesListPreference!!.isVisible = showFajrDegreeSelector
        ishaDegreesListPreference!!.isVisible = showIshaDegreeSelector
        apiSettingsPreferenceCategory!!.isVisible =
            fajrDegreesListPreference!!.isVisible || ishaDegreesListPreference!!.isVisible
    }

    // endregion methods
    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(
            _preferenceChangeListener
        )
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(
            _preferenceChangeListener
        )
    }

    companion object {
        // region static fields
        private const val PREFERENCE_CATEGORY_API_SETTINGS = "apiSettingsPreferenceCategory"
        private const val PREFERENCE_NAME_API_SELECTION = "apiSelection"
        private const val PREFERENCE_NAME_MINUTE_ADJUSTMENT_SELECTION = "minuteAdjustmentSelection"
        private const val PREFERENCE_NAME_FAJR_DEGREE_SELECTION = "fajrCalculationDegree"
        private const val PREFERENCE_NAME_ISHA_DEGREE_SELECTION = "ishaCalculationDegree"
    }

    // endregion fields
    init {
        prayerTypeWithMomentType = AbstractMap.SimpleEntry(
            prayerType,
            if (_isBeginning) EPrayerTimeMomentType.Beginning else EPrayerTimeMomentType.End
        )
    }
}